### PR TITLE
Add prefer_v2_templates parameter to Reindex

### PR DESCRIPTION
**Pull request by probakowski**
Relates to #53101

### CODE REVIEW

**Comment by probakowski** [comment]
@elasticmachine update branch

**Comment by probakowski** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/index/reindex/ReindexRequest.java [hunk 1]
```diff
@@ -86,6 +91,9 @@ public ReindexRequest(StreamInput in) throws IOException {
         super(in);
         destination = new IndexRequest(in);
         remoteInfo = in.readOptionalWriteable(RemoteInfo::new);
+        if (in.getVersion().onOrAfter(Version.V_7_8_0)) {
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/index/reindex/ReindexRequest.java:94]: I think this is going to have to be 8.0.0 until this gets backported

File: server/src/main/java/org/elasticsearch/index/reindex/ReindexRequest.java [hunk 2]
```diff
@@ -262,6 +280,9 @@ public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
         destination.writeTo(out);
         out.writeOptionalWriteable(remoteInfo);
+        if (out.getVersion().onOrAfter(Version.V_7_8_0)) {
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/index/reindex/ReindexRequest.java:283]: Same comment here about this having to be 8.0.0 until backport (or else the BWC tests will likely fail)

**Review by dakrone — APPROVED** [review]
Thanks for doing this @probakowski! I left two really minor comments that are just related to BWC for the request, otherwise this looks good to me!

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java [hunk 3]
```diff
@@ -68,6 +68,11 @@ protected ReindexRequest buildRequest(RestRequest request) throws IOException {
         if (request.hasParam("scroll")) {
             internal.setScroll(parseTimeValue(request.param("scroll"), "scroll"));
         }
+
+        if (request.hasParam("prefer_v2_templates")) {
+            internal.preferV2Templates(Boolean.parseBoolean(request.param("prefer_v2_templates")));
```
Comment (andreidan) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java:73]: for 8.0 we want this to default to `true` if it's not specified by the user (this is just a reminder to change this after backport)

**Review by andreidan — APPROVED** [review]
LGTM, thanks Przemko

File: server/src/main/java/org/elasticsearch/index/reindex/ReindexRequest.java (hunk 1 above)
Comment (probakowski) [reply server/src/main/java/org/elasticsearch/index/reindex/ReindexRequest.java]: Yep, you're right, changed it and will follow up after backport

File: server/src/main/java/org/elasticsearch/index/reindex/ReindexRequest.java (hunk 2 above)
Comment (probakowski) [reply server/src/main/java/org/elasticsearch/index/reindex/ReindexRequest.java]: Yep, you're right, changed it and will follow up after backport

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java (hunk 3 above)
Comment (probakowski) [reply modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java:73]: I wonder if we should change it here or simply pass null down the road and let create index request pick up correct default (for making it DRY and using only single place to set defaults)

**Comment by probakowski** [comment]
@elasticmachine run elasticsearch-ci/2

**Pull request merged**